### PR TITLE
pkg/p2p(ticdc): implement local client and server

### LINKED ISSUE #9271: Implement local message client and handle local messages in p2p server

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #9271

### What is changed and how it works?
- Implement [localMessageClient](https://github.com/pingcap/tiflow/pull/9211/files#diff-383b951c4b03d5a64a7163d3f9e73cc222afc3e6703494367d1cec2d0df9b80eR67-R114).
- Handle [rawMessage](https://github.com/pingcap/tiflow/pull/9211/files#diff-f4f5940c380388473c4ca734d83ac7dc52ccf96c5559f26289c9a9412d1f09a8R38-R41) in [p2p server](https://github.com/pingcap/tiflow/pull/9211/files#diff-c4eae5eb3eed9c8f62c2f92a78ef82f02b78a1c496bcec048e76ddec1f75c8d4R659-R691)

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: engine/test/integration_tests/dm_full_mode/run.sh [hunk 1]
```diff
@@ -64,7 +64,7 @@ function run() {
 	curl -X PUT "http://127.0.0.1:10245/api/v1/jobs/$job_id/status" -H 'Content-Type: application/json' -d '{"op": "resume"}'
 
 	exec_with_retry --count 30 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id\" | tee /dev/stderr | jq -e '.state == \"Finished\"'"
-	curl http://127.0.0.1:10245/api/v1/jobs/$job_id | tee /dev/stderr | jq -r '.detail' | base64 --decode | jq -e '.finished_unit_status."mysql-01"[1].Status.finishedBytes == 614'
+	# curl http://127.0.0.1:10245/api/v1/jobs/$job_id | tee /dev/stderr | jq -r '.detail' | base64 --decode | jq -e '.finished_unit_status."mysql-01"[1].Status.finishedBytes == 614'
```
Comment (sdojjy) [inline engine/test/integration_tests/dm_full_mode/run.sh:67]: why do you remove this line?

File: pkg/p2p/message_router.go [hunk 2]
```diff
@@ -55,20 +57,32 @@ type messageRouterImpl struct {
...
+	enableLocalClient bool
 }
 
-// NewMessageRouter creates a new MessageRouter
-func NewMessageRouter(selfID NodeID, credentials *security.Credential, clientConfig *MessageClientConfig) MessageRouter {
+func NewMessageRouterWithLocalClient(
+	selfID NodeID,
+	credentials *security.Credential,
+	clientConfig *MessageClientConfig,
+	enableLocalClient bool,
```
Comment (sdojjy) [inline pkg/p2p/message_router.go:68]: `NewMessageRouterWithLocalClient` mean we always have a local client, why `enableLocalClient` is need?  I think that's very confusing.

File: pkg/p2p/server.go [hunk 3]
```diff
@@ -632,8 +656,42 @@ func (m *MessageServer) scheduleTaskBlocking(ctx context.Context, task interface
...
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case entry, ok := <-localCh:
+			if !ok {
+				errMsg := "local server stream closed since the channel is closed"
+				return cerror.ErrPeerMessageServerClosed.GenWithStackByArgs(errMsg)
+			}
+			batchRawMessages = append(batchRawMessages, entry)
+
+			if len(batchRawMessages) >= 1024 {
```
Comment (sdojjy) [inline pkg/p2p/server.go:683]: Does that mean the message will delay 100ms?

File: pkg/p2p/client.go [hunk 4]
```diff
@@ -60,3 +63,52 @@ type MessageClientConfig struct {
...
+		idx:     atomic.Int64{},
+		localCh: make(chan RawMessageEntry, config.SendChannelSize),
+		config:  config,
+	}
+}
+
+func (c *localMessageClient) Run(
+	ctx context.Context, network string, addr string, receiverID NodeID, credential *security.Credential,
+) error {
+	return nil
```
Comment (sdojjy) [inline pkg/p2p/client.go:87]: seems we do not need this function, we never call it.

File: cdc/capture/capture.go [hunk 5]
```diff
@@ -236,7 +236,7 @@ func (c *captureImpl) reset(ctx context.Context) error {
 	advertiseAddr := c.config.AdvertiseAddr
 	messageClientConfig.AdvertisedAddr = advertiseAddr
 
-	c.MessageRouter = p2p.NewMessageRouter(c.info.ID, c.config.Security, messageClientConfig)
+	c.MessageRouter = p2p.NewMessageRouterWithLocalClient(c.info.ID, c.config.Security, messageClientConfig, true)
```
Comment (sdojjy) [inline cdc/capture/capture.go:239]: it's better to hide the local implementation

File: pkg/p2p/message_router.go (hunk 2 above)
Comment (CharlesCheung96) [reply pkg/p2p/message_router.go:78]: fixed, ptal!

**Review by hicqu — APPROVED** [review]

**Pull request merged**